src/components/modals/FormStepSaveModal.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Display a modal to allow the user to save the form step in it's current state.
+ */
+import React, {useContext} from 'react';
+import PropTypes from 'prop-types';
+import {useHistory} from 'react-router-dom';
+import {useIntl, FormattedMessage} from 'react-intl';
+
+import {put, post, destroy} from 'api';
+import { ConfigContext } from 'Context';
+import Body from 'components/Body';
+import ErrorMessage from 'components/ErrorMessage';
+import HelpText from 'components/HelpText';
+import Label from 'components/Label';
+import Modal from 'components/modals/Modal';
+import Input from 'components/Input';
+import {Toolbar, ToolbarList} from 'components/Toolbar';
+import Button from 'components/Button';
+import {getBEMClassName} from 'utils';
+import {useImmerReducer} from "use-immer";
+import Loader from "../Loader";
+
+
+const initialState = {
+  email: "",
+  errorMessage: "",
+  isSaving: false,
+};
+
+const reducer = (draft, action) => {
+  switch(action.type) {
+    case 'SET_EMAIL': {
+      draft.email = action.payload;
+      break;
+    }
+    case 'SET_ERROR_MESSAGE': {
+      draft.errorMessage = action.payload;
+      break;
+    }
+    case 'SET_IS_SAVING_TRUE': {
+      draft.isSaving = true;
+      break;
+    }
+    case 'SET_IS_SAVING_FALSE': {
+      draft.isSaving = false;
+      break;
+    }
+    default: {
+      throw new Error(`Unknown action ${action.type}`);
+    }
+  }
+};
+
+
+const FormStepSaveModal = ({
+    isOpen,
+    closeModal,
+    stepData,
+    saveStepDataUrl,
+    suspendFormUrl,
+}) => {
+  const history = useHistory();
+  const intl = useIntl();
+  const config = useContext(ConfigContext);
+
+  // const [email, setEmail] = useState("");
+  // const [errorMessage, setErrorMessage] = useState("");
+
+  const [
+    {email, errorMessage, isSaving},
+    dispatch
+  ] = useImmerReducer(reducer, initialState);
+
+  const onSubmit = async (event) => {
+    event.preventDefault();
+    dispatch({type: 'SET_IS_SAVING_TRUE'});
+    dispatch({type: 'SET_ERROR_MESSAGE', payload: ''});
+    let response = await put(saveStepDataUrl, {data: stepData});
+    if (!response.ok) {
+      dispatch({
+        type: 'SET_ERROR_MESSAGE',
```

Comment:
here too - would use some sort of action type like `API_ERROR` and keep the payload. That the error message state is set is not semantically interesting, it's just a side effect, an implementation detail.